spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java

```diff
@@ -78,8 +81,20 @@ S3ClientBuilder s3ClientBuilder(AwsClientBuilderConfigurer awsClientBuilderConfi
 	@ConditionalOnMissingBean(S3Operations.class)
 	@ConditionalOnBean(S3ObjectConverter.class)
 	S3Template s3Template(S3Client s3Client, S3OutputStreamProvider s3OutputStreamProvider,
-			S3ObjectConverter s3ObjectConverter) {
-		return new S3Template(s3Client, s3OutputStreamProvider, s3ObjectConverter);
+			S3ObjectConverter s3ObjectConverter, S3Presigner s3Presigner) {
+		return new S3Template(s3Client, s3OutputStreamProvider, s3ObjectConverter, s3Presigner);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	S3Presigner s3Presigner(S3Properties properties, AwsProperties awsProperties,
```

Comment:
Update `S3AutoConfigurationTests` to test if this bean is created correctly with different properties.

Reply:
You would need to assert on the actual `S3Presigner` bean - but as you have probably noticed it does not expose methods to get the endpoint and other properties. We had the same issue with other AWS clients and solved it by creating a concept of [ConfiguredAwsClient](https://github.com/awspring/spring-cloud-aws/blob/main/spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/ConfiguredAwsClient.java) that use reflection to access properties that are interesting for tests. Take a look how it is used in `S3AutoConfigurationTests`: https://github.com/awspring/spring-cloud-aws/blob/main/spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfigurationTests.java#L147

Of course it is not ideal since the test can start failing if internals of how AWS SDK works change, but it is definitely better than not testing at all or switching to integration tests that enable testing such properties.

Reply:
I did notice that, but I don't think I can use ConfiguredAwsClient because it takes an SdkClient whereas S3Presigner inherits from SdkPresigner. Should I make a separate version called ConfiguredAwsPresigner that takes an SdkPresigner? Alternatively, since SdkPresigner and SdkClient both inherit from SdkAutoCloseable I could try changing ConfiguredAwsClient to take an SdkAutoCloseable.

Reply:
Whatever makes sense - I would need to go through internals of each to see if they have same structure so that reflections work on them. Gut feeling is that making `ConfiguredS3Presigner` makes the most sense.